### ISSUE TITLE
Wrong music label (32-bit)

### ISSUE THREAD

**Issue by KOPRajs** [xbmc/xbmc#22355]
### Describe the bug

While playing typical 16-bit, 44 kHz, 320 kbps MP3 file Kodi shows 32-bit label instead of 16-bit.

### To Reproduce

1. Play 16-bit (320 kbps) MP3 file
2. Go to fullscreen
3. Press "i" to show information and see the 32-bit label in the bottom right corner

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [ ] tvOS
 - [X] Linux
 - [ ] macOS
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: CoreELEC 20.0 RC2
 - Kodi version: 20.0 RC2

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by enen92** [comment]
I doubt there's anything to fix here:

>> Bit depth is only meaningful in reference to a PCM digital signal. Non-PCM formats, such as lossy compression formats, do not have associated bit depths. For example, in MP3, quantization is performed on PCM samples that have been transformed into the frequency domain.

ffprobe doesn't report anything for me:

```
codec_name=mp3
codec_long_name=MP3 (MPEG audio layer 3)
profile=unknown
codec_type=audio
codec_tag_string=[0][0][0][0]
codec_tag=0x0000
sample_fmt=fltp
sample_rate=44100
channels=2
channel_layout=stereo
bits_per_sample=0

```

Source: https://superuser.com/questions/553552/how-to-determine-aac-bit-depth

@fritsch might know way more about this though

**Comment by KOPRajs** [comment]
> Bit depth is only meaningful in reference to a PCM digital signal.

Right, then why Kodi shows 32-bit label? Regular CD, from which most mp3 files come from, is 16-bit PCM, most audio chips can output maximum up to 24-bit PCM, so showing 32-bit label for mp3 doesn't seem right.

**Comment by the-black-eagle** [comment]
Ok, so.....as far as I can remember, although @fritsch is by far the best guy to know, I _think_ that what gets reported is the output stream.  So, Kodi reports what it opens as output, not as input and a 16bit input stream gets padded to a 32bit output stream.  No loss of quality there, just a bigger number with more zeroes in it.

Could be wrong, and I'll take the hit if I am, but I think that's how it works.....

**Comment by enen92** [comment]
Exactly, the 32bit value comes from here: https://github.com/xbmc/xbmc/blob/50c9d4263c7c2ea071b46c2a039cc78626cd050c/xbmc/cores/AudioEngine/Utils/AEUtil.cpp#L121

**Comment by fritsch** [comment]
The intermediate format of Audio engine for PCM is planar float with a 32 bit precission. MP3 needs to be decoded before you can play it on your soundcard. In order to not loose anything during this decoding, we use the maximum internal format available (floatp). From here it is mapped to a format the sink supports (16 bit int, 32 bit int or 24 bit int inside 32 bit ... and so on).

So what kodi shows here is not really the input format, but the format kodi has after decoding.

**Comment by KOPRajs** [comment]
I see, the question is if this is a good design then. Does it make sense to show users an internal maximum bit depth that doesn't actually depends on the file itself? Maybe move the bit depth calculation somewhere else and only set it for the formats where id does make sense?

**Comment by fritsch** [comment]
Good question. If you press "i" do you expect to see what the "file is on disk" or "what you are currently listening", e.g. the output of the decoder?

I can find arguments for both. Send a PR to whatever you like best I'd say then we can discuss it there. Thanks much in advance.

**Comment by KOPRajs** [comment]
I'm not sure either, but the concept of the labels (at least as I understand it) is about the files on disk. Otherwise the rest of the labels doesn't make sense (MP3, 320kbps).

**Comment by fritsch** [comment]
Yes, that is right. It's at least confusing.

**Comment by wutschel** [comment]
I stumbled across the same when trying to detect when a HiRes audio is playing and decided to ignore all lossy codecs (Kodi provides 32-bit as sample resolution for those).

When thinking through what a lossy codec does and what it represents, you cannot really tell which sample resolution (or even sample rate) the source had as this information is not present in the bitstream. Such codecs -- also the aged MP1/2/3 -- are in principle able to encode signals with higher resolution as 16 bit as they internally use a float representation as well. The origin might have been using even 24 bit. The _quality_ as we all know is defined by the codec, the overall bitrate and the codec's settings while encoding.

To summarize: With lossy codecs you simply do not know the source sample resolution. You can either just not show it, or display the maximum you can get out of the stream -- which is the sink's sample resolution, 32 bit in many cases. All other solutions (e.g. "16 bit" for MP3) or just guesses, which are also potentially wrong, especially with the increasing availability of HiRes source material.

So, if you _must_ show a sample resolution (for UI reasons), "32 bit" is the best compromise from my point if view.

**Comment by KOPRajs** [comment]
I agree, the best solution would be to not show bit resolution at all if it does not make sense (e.g. for lossy codecs).
